app/models/miq_worker.rb

```diff
@@ -128,8 +128,30 @@ def self.enough_resource_to_start_worker?
     MiqServer.my_server.enough_resource_to_start_worker?(self)
   end
 
+  def self.cleanup_failed_workers
+    if containerized_worker?
+    elsif systemd_worker?
+      failed_services.each do |service|
+        cleanup_systemd_worker(service)
+      end
+    end
+  end
+
+  def self.current_workers
+    if containerized_worker?
+      # TODO search container deployments
+      include_stopping_workers_on_synchronize ? find_alive : find_current_or_starting
+    elsif systemd_worker?
+      active_services
+    else
+      include_stopping_workers_on_synchronize ? find_alive : find_current_or_starting
+    end
+  end
+
   def self.sync_workers
-    w       = include_stopping_workers_on_synchronize ? find_alive : find_current_or_starting
+    cleanup_failed_workers
```

Comment:
Is this expensive to do every run through sync_workers?

Reply:
I ask this because I was looking at talking to kubernetes and it seems wasteful to go worker class by worker class looking for failed workers/deployments for each type and using kubeclient to ask for each and every worker class.  It feels like we did this class by class because we thought DB queries were cheap but we could fetch the current workers once in `MiqServer#monitor_workers` before it calls `sync_workers`.  

Short term: the `cleanup_failed_workers` could be called from monitor_workers if systemd/containers provided we have a way to label and search for them in systemd/containers, which is available with containers using labels.  `cleanup_failed_workers` method could call `cleanup_failed_deployments` in the case of containers.


```ruby
class ContainerOrchestrator
  def get_pods(options)
    options = {:namespace => my_namespace}.merge(options)
    kube_apps_connection.get_pods(options)
  end
end
```

```ruby
module ContainerCommon  # included into Miqworker
  def failed_deployments(restart_count = 5)
    ContainerOrchestrator.new.get_pods(:label_selector => "app=manageiq,worker=true").collect do |pod|
      pod.status.containerStatuses.select {|cs| cs.lastState.terminated && cs.restartCount > restart_count }
    end.flatten
  end

  def cleanup_failed_deployments
    failed_deployments.each do |failed|
      ContainerOrchestrator.new.delete_deployment(failed.name)
    end
  end
end
```

If need be, we can add another label to workers to indicate it's an ems based worker so we can search for them and possibly only handle ems workers in this way to start.  Maybe once we implement this basic functionality, we can do other worker types.

Reply:
I ask this because I was looking at talking to kubernetes and it seems wasteful to go worker class by worker class looking for failed workers/deployments for each type and using kubeclient to ask for each and every worker class.  It feels like we did this class by class because we thought DB queries were cheap but we could fetch the current workers once in `MiqServer#monitor_workers` before it calls `sync_workers`.  

Short term: the `cleanup_failed_workers` could be called from monitor_workers if systemd/containers provided we have a way to label and search for them in systemd/containers, which is available with containers using labels.  `cleanup_failed_workers` method could call `cleanup_failed_deployments` in the case of containers.


```ruby
class ContainerOrchestrator
  def get_pods(options)
    options = {:namespace => my_namespace}.merge(options)
    kube_apps_connection.get_pods(options)
  end
end
```

```ruby
module ContainerCommon  # included into Miqworker
  def failed_deployments(restart_count = 5)
    ContainerOrchestrator.new.get_pods(:label_selector => "app=manageiq,worker=true").collect do |pod|
      pod.status.containerStatuses.select {|cs| cs.lastState.terminated && cs.restartCount > restart_count }
    end.flatten
  end

  def cleanup_failed_deployments
    failed_deployments.each do |failed|
      ContainerOrchestrator.new.delete_deployment(failed.name)
    end
  end
end
```

If need be, we can add another label to workers to indicate it's an ems based worker so we can search for them and possibly only handle ems workers in this way to start.  Maybe once we implement this basic functionality, we can do other worker types.

Reply:
Ultimately I think we should use a kubernetes watch, and keep an in memory inventory of the workers that the watch updates directly.  Then it's not expensive to loop through the in-memory values at all, even if done worker-by-worker.  Even so, if we have all of the data, then I agree there's no _need_ to go worker by worker.

Reply:
Yeah, I was looking at watches.  I'll need to recreate the constant restart failures to see how they're represented and to see if it's possible to miss any events and have to play "catch up".  I was thinking of the above code change for possible backport whereas we might do watches or "additional changes such as labelling ems based workers metadata so we can search by those labels" on master.  Note, if we're going to get the data for all workers anyway, labels for ems based workers or other more specific details would not be needed as labels if we're not going to be querying using them but instead annotations could be used.

Reply:
FYI, for backport or just a minor change, the only other thing we'd need to change is to add the label on the workers `workers=true` or something.

For the longer term change with watches or a combination of watches and scheduled "every 10 minute" manual check using the `get_pods` code above or just the `get_pods` above, we can add more labelling and annotations on creation and then use them on querying to make our systemd/pods code for looking up current/failed be as similar as possible.

Reply:
Hm no way to "label" a service that I know of but I can make something work based off of the worker names, agree being able to do this from MiqServer not each worker would be better

Reply:
Also :+1: to k8s watches, happy to help you with that we've used them extensively in k8s/ocp refresh
You get a resource_version back for a collection so you won't "miss" events, you just ask for events since the last time you saw them.  They aren't kept around forever so if you don't ask for a looong time worst case you get back a 410 Gone error and you have to do the full request again.

Reply:
For the purposes of this PR, I think this is resolved.  I'll be adding a PR to discuss how we talk to kubernetes using watches and if we want to add labels to make searching for the pods we want to manage (not orchestrator or pg, etc.)